graph/telemetry/istio/util/util.go

```diff
@@ -157,3 +162,62 @@ func GetApp(rates graph.RequestedRates) string {
 	}
 	return ""
 }
+
+// PromQuery queries Prometheus for metric data
+func PromQuery(ctx context.Context, query string, queryTime time.Time, api prom_v1.API, conf *config.Config) model.Vector {
+	return PromQueryAppender(ctx, query, queryTime, api, conf, nil)
+}
+
+// PromQueryAppender is for appenders to query Prometheus for metric data
+func PromQueryAppender(ctx context.Context, query string, queryTime time.Time, api prom_v1.API, conf *config.Config, a graph.Appender) model.Vector {
+	if query == "" {
+		return model.Vector{}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// get logger from context
+	zl := log.FromContext(ctx)
+
+	// add scope if necessary
+	query = AddQueryScope(query, conf)
+
+	// wrap with a round() to be in line with metrics api
+	query = fmt.Sprintf("round(%s,0.001)", query)
+
+	// start our timer
+	var promtimer *prom_client.Timer
+	if a == nil {
+		promtimer = internalmetrics.GetPrometheusProcessingTimePrometheusTimer("Graph-Generation")
+	} else {
+		promtimer = internalmetrics.GetPrometheusProcessingTimePrometheusTimer("Graph-Appender-" + a.Name())
+	}
+
+	// perform the Prometheus query now
+	value, warnings, err := api.Query(ctx, query, queryTime)
+
+	// log warnings and abort immediately on errors
+	if len(warnings) > 0 {
+		zl.Warn().Str("problemQuery", query).Msgf("PromQuery: Prometheus Warnings: [%s]", strings.Join(warnings, ","))
+	}
+	if err != nil {
+		zl.Trace().Str("failedQuery", query).Msgf("PromQuery: Prometheus Error: [%v]", err)
+	}
+	graph.CheckUnavailable(err)
+
+	// notice we only collect metrics and log a message for successful prom queries
+	duration := promtimer.ObserveDuration()
```

Comment:
Does this still work if you don't have the internal metrics enabled? I'm not sure if you can even disable those.

Reply:
Yes. Internal metric enablement only turns on or off the server endpoint (see [here](https://github.com/kiali/kiali/blob/v2.9.0/server/server.go#L148-L151)) - all the metrics are still collected